### PR TITLE
Enhancement: Replace Starlette Responses

### PR DESCRIPTION
**Pull request by Goldziher**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

This PR removes the usage of Starlette based `Response` objects by introducing our own version - with a compatible API. 

Relates to #612

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -70,12 +70,10 @@ repos:
             "flake8-print",
             "flake8-simplify",
             "flake8-type-checking",
-            "flake8-pytest-style",
             "flake8-implicit-str-concat",
             "flake8-noqa",
             "flake8-return",
             "flake8-secure-coding-standard",
-            "flake8-eradicate",
```
Comment (Goldziher) [inline .pre-commit-config.yaml:78]: this plugin was giving false positives on comments

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -70,12 +70,10 @@ repos:
             "flake8-print",
             "flake8-simplify",
             "flake8-type-checking",
-            "flake8-pytest-style",
```
Comment (Goldziher) [inline .pre-commit-config.yaml:73]: this plugin was just noise IMO

**Comment by Goldziher** [comment]
PR is still missing docs and some polishing touches, but it is substantially done and the tests are in place - so its a good time to start reviewing. Please do!

**Review by peterschutt — COMMENTED** [review]
Haven't made all the way through yet!

File: starlite/datastructures/response_containers.py [hunk 3]
```diff
@@ -237,6 +236,9 @@ def to_response(
         Returns:
             A TemplateResponse instance
         """
+        from starlite.response import (  # pylint: disable=import-outside-toplevel
+            TemplateResponse,
+        )
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:241]: apparently its resolved!

**Comment by infohash** [comment]
So with this PR, instead of using starlette `RedirectResponse` in the middleware, I will be able to use `RedirectResponse` from starlite?

**Comment by Goldziher** [comment]
> So with this PR, instead of using starlette `RedirectResponse` in the middleware, I will be able to use `RedirectResponse` from starlite?

that is correct

File: starlite/response/base.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        Raises:
+            TypeError: if value is not supported
+        """
+        return default_serializer(value)
+
+    def render(self, content: Any) -> bytes:
+        """
+        Handles the rendering of content T into a bytes string.
+        Args:
+            content: An arbitrary value of type T
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/base.py:201]: fixed

**Comment by provinzkraut** [comment]
[In docs/usage/5-responses/2-status-codes.md](https://github.com/starlite-api/starlite/blob/replace-response/docs/usage/5-responses/2-status-codes.md?plain=1#L40) this line should be:

```diff
- tests). Starlette includes easy to use statuses that are exported from `starlette.status`, e.g. `HTTP_200_OK`
+ tests). Starleite includes easy to use statuses that are exported from `starlite.status`, e.g. `HTTP_200_OK`
```

**Comment by Goldziher** [comment]
> [In docs/usage/5-responses/2-status-codes.md](https://github.com/starlite-api/starlite/blob/replace-response/docs/usage/5-responses/2-status-codes.md?plain=1#L40) this line should be:
> 
> ```diff
> - tests). Starlette includes easy to use statuses that are exported from `starlette.status`, e.g. `HTTP_200_OK`
> + tests). Starleite includes easy to use statuses that are exported from `starlite.status_codes`, e.g. `HTTP_200_OK`
> ```

fixed

**Review by provinzkraut — APPROVED** [review]

File: starlite/datastructures/response_containers.py [hunk 5]
```diff
@@ -24,21 +24,24 @@
...
+    StreamingResponse,
+    TemplateResponse,
+)
 
 if TYPE_CHECKING:
     from starlite.app import Starlite
     from starlite.connection import Request
 
-R = TypeVar("R", bound=StarletteResponse)
+R = TypeVar("R")
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:43]: I agree

File: starlite/datastructures/response_containers.py [hunk 6]
```diff
@@ -57,6 +59,8 @@ class Config(BaseConfig):
     """A list of Cookie instances to be set under the response 'Set-Cookie' header. Defaults to None."""
     media_type: Optional[Union[MediaType, str]] = None
     """If defined, overrides the media type configured in the route decorator"""
+    encoding: str = "utf-8"
+    """The encoding to used for the response headers."""
```
Comment (jtraub) [inline starlite/datastructures/response_containers.py:63]: Probably this should be `The encoding to be used for the response headers`

File: starlite/datastructures/response_containers.py [hunk 7]
```diff
@@ -88,22 +92,36 @@ class File(ResponseContainer[FileResponse]):
...
+    filename: Optional[str] = None
     """The filename"""
     stat_result: Optional[os.stat_result] = None
     """File statistics"""
+    chunk_size: int = ONE_MEGA_BYTE
+    """The size of chunks to use when streaming the file"""
+    content_disposition_type: "Literal['attachment', 'inline']" = "attachment"
+    """The type of the 'Content-Disposition'. Either 'inline' or 'attachment'."""
+    etag: Optional[str] = None
+    """An optional etag for the file. If not provided, and etag will be generated automatically."""
```
Comment (jtraub) [inline starlite/datastructures/response_containers.py:104]: Another typo? `If not provided, an etag`

File: starlite/datastructures/response_containers.py (hunk 6 above)
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:63]: fixed

File: starlite/datastructures/response_containers.py (hunk 7 above)
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:104]: fixed

File: starlite/datastructures/response_containers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:43]: I bound it to `Response` - before i had some cyclical import issues, but thats fixed now.

[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/datastructures/response_containers.py]: Ok, I remembered why I did this. It shouldn't be a protocol - it should be `Response`, but mypy is going ape shit for some reason. I have no clue why to be honest, i tried fixing it but its some issue with GenericModel + Generic and mypy in this setup. Anyhow, for now I will keep it without binding, because we are using ResponseContainer generics internally.

File: starlite/datastructures/cookie.py [hunk 8]
```diff
@@ -51,3 +51,18 @@ def to_header(self, **kwargs: Any) -> str:
             if value is not None:
                 simple_cookie[self.key][key] = value
         return simple_cookie.output(**kwargs).strip()
+
+    def __eq__(self, other: Any) -> bool:
```
Comment (provinzkraut) [inline starlite/datastructures/cookie.py:55]: I'm actually not sure if this is a good approach. It doesn't really do a strict equality check, since it does not compare for value and other attributes. Maybe this shouldn't be a dunder method, but a separate `is_equal` method with a `strict` flag defaulting to `True`. We could still implement `__eq__` as a callback to  `self.is_equal(strict=True)` then.

I'd also do the canonical "return `NotImplemented`" if the types don't match

File: starlite/response/base.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        *,
+        status_code: int = HTTP_200_OK,
+        media_type: Union[MediaType, "OpenAPIMediaType", str] = MediaType.JSON,
+        background: Optional[Union["BackgroundTask", "BackgroundTasks"]] = None,
+        headers: Optional[Dict[str, Any]] = None,
+        cookies: Optional["ResponseCookies"] = None,
+        encoding: str = "utf-8",
+        is_head_response: bool = False,
+    ) -> None:
+        """This is the base Starlite HTTP response class, its is used as a
```
Comment (provinzkraut) [inline starlite/response/base.py:70]: Typo: "its is used" > "it is used"
Typo: "Used as a basis" > "Used as the basis"

File: tests/response/test_base_response.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    @get("/delete")
+    def delete_cookie_handler() -> Response:
+        response = Response(content=None)
+        response.delete_cookie(
+            "test",
+            "abc",
+        )
+        assert len(response.cookies) == 1
+        return response
+
```
Comment (provinzkraut) [inline tests/response/test_base_response.py:82]: Since values don't matter for cookie deletion, maybe add a test for that case. I.e. that a cookie is deleted if its key matches but its value does not.

File: tests/response/test_base_response.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        return Response(content=body, status_code=status)
+
+    with create_test_client(handler) as client:
+        response = client.get("/")
+        if should_raise:
+            assert response.status_code == HTTP_500_INTERNAL_SERVER_ERROR
+        else:
+            assert response.status_code == status
+            assert "content-length" not in response.headers
+
```
Comment (provinzkraut) [inline tests/response/test_base_response.py:139]: Maybe add a test for the case that a handler **should** return `None`, but doesn't (and the other way around) when no explicit status code is set (i.e a handler is annotated as `- > None`  without a return code)

File: tests/response/test_file_response.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    assert filled_by_bg_task == ""
+    client = TestClient(app)
+    response = client.get("/")
+    expected_disposition = 'attachment; filename="example.png"'
+    assert response.status_code == HTTP_200_OK
+    assert response.content == content
+    assert response.headers["content-type"] == "image/png"
+    assert response.headers["content-disposition"] == expected_disposition
+    assert "content-length" in response.headers
+    assert "last-modified" in response.headers
```
Comment (provinzkraut) [inline tests/response/test_file_response.py:56]: Probably best to test the correctness of these values

File: tests/response/test_streaming_response.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+            yield b"chunk "
+
+    response = StreamingResponse(content=stream_indefinitely())
+
+    with anyio.move_on_after(1) as cancel_scope:
+        await response({}, receive_disconnect, send)  # type: ignore
+    assert not cancel_scope.cancel_called, "Content streaming should stop itself."
+
+
+def test_streaming_response() -> None:
```
Comment (provinzkraut) [inline tests/response/test_streaming_response.py:62]: This should be tested with a synchronous generator as well 
(also applies to the following tests)

**Review by provinzkraut — CHANGES_REQUESTED** [review]

File: starlite/response/base.py [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+        Args:
+            key: Header key.
+            value: Header value.
+
+        Returns:
+            None.
+        """
+        self.headers[key] = value
+
+    def set_etag(self, etag: str) -> None:
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/base.py:151]: ok

File: starlite/response/base.py [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+        return b""
+
+    @property
+    def content_length(self) -> Optional[int]:
+        """
+
+        Returns:
+            Returns the value for the 'Content-Length' header, if applies.
+        """
+        if self.status_allows_body and isinstance(self.body, bytes):
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/base.py:240]: fair point

File: starlite/response/base.py [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+            return len(self.body)
+        return None
+
+    @property
+    def encoded_headers(self) -> List[Tuple[bytes, bytes]]:
+        """
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/base.py:246]: done

File: starlite/response/file.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+        """Creates an etag.
+
+        Notes:
+            - Function is derived from flask.
+
+        Returns:
+            An etag.
+        """
+        check = adler32(str(path).encode("utf-8")) & 0xFFFFFFFF
+        return f"{self.stat_result.st_mtime}-{self.stat_result.st_size}-{check}"
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/file.py:123]: done

File: starlite/response/redirect.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+            status_code: An HTTP status code. The status code should be one of 301, 302, 303, 307 or 308,
+                otherwise an exception will be raised. .
+            headers: A string keyed dictionary of response headers. Header keys are insensitive.
+            cookies: A list of [Cookie][starlite.datastructures.Cookie] instances to be set under the response 'Set-Cookie' header.
+            encoding: The encoding to used for the response headers.
+        Raises:
+            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If status code is not a redirect status code.
+        """
+        if status_code not in REDIRECT_STATUS_CODES:
+            raise ImproperlyConfiguredException(
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/redirect.py:41]: Done

File: starlite/response/base.py [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+
+        Args:
+            key: Key of the cookie.
+            path: Path of the cookie.
+            domain: Domain of the cookie.
+
+        Returns:
+            None.
+        """
+        self.cookies = [
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/response/base.py:178]: nice

File: starlite/datastructures/cookie.py (hunk 8 above)
Comment (Goldziher) [reply starlite/datastructures/cookie.py:55]: cookie equality is based on these parameters in browsers, thats the reason. Why wouldnt this be a good idea in your opinion?

File: starlite/response/base.py (hunk 9 above)
Comment (Goldziher) [reply starlite/response/base.py:70]: fixed

File: tests/response/test_streaming_response.py (hunk 13 above)
Comment (Goldziher) [reply tests/response/test_streaming_response.py:62]: there are tests with sync generators elsewhere that use this, but i will add

File: tests/response/test_base_response.py (hunk 10 above)
Comment (Goldziher) [reply tests/response/test_base_response.py:82]: can do. The point really is what makes the browser delete the cookie, which is not the value but rather the key+path+domain combination.

File: tests/response/test_base_response.py (hunk 11 above)
Comment (Goldziher) [reply tests/response/test_base_response.py:139]: aight

File: tests/response/test_file_response.py (hunk 12 above)
Comment (Goldziher) [reply tests/response/test_file_response.py:56]: ok

Comment (Goldziher) [reply tests/response/test_file_response.py:56]: done

File: starlite/datastructures/cookie.py (hunk 8 above)
Comment (provinzkraut) [reply starlite/datastructures/cookie.py:55]: It makes sense in the HTTP world, but less so in the python world. In the python world it might be expected that two cookies are only equal if their public attributes all equal.

You have a point though, for this specific implementation it makes sense to check equality like this. Probably should be mentioned in the docs somewhere then.

File: tests/response/test_base_response.py (hunk 10 above)
Comment (provinzkraut) [reply tests/response/test_base_response.py:82]: Correct. I just think we should have a test that checks that we are doing it on our side as well. What I mean is:

Test that `response.delete_cookie("test, "abc")` should deletes any cookie that matches these parameters, but ignore things like its value.

File: tests/response/test_streaming_response.py (hunk 13 above)
Comment (provinzkraut) [reply tests/response/test_streaming_response.py:62]: Must have missed them! Maybe move them here then because that's where you'd expect them (at least I did)

**Review by provinzkraut — APPROVED** [review]
